tests/test_result_do.py

```diff
@@ -24,25 +24,179 @@ def _get_output(is_suc1: bool, is_suc2: bool) -> Result[float, int]:
+    assert await _aget_output(False, False, False) == Err(1)
+
+    async def _aget_output_return_immediately(
+        is_suc1: bool, is_suc2: bool, is_suc3: bool
+    ) -> Result[float, int]:
+        return await do_async(
+            Ok(len(x) + int(y) + z)
+            for x in await aget_resx(is_suc1)
+            for y in await aget_resy(is_suc2)
+            for z in get_resz(is_suc3)
+        )
+
+    assert await _aget_output_return_immediately(True, True, True) == Ok(6.5)
+
+
+@pytest.mark.asyncio
+async def test_result_do_async_further_processing() -> None:
+    async def aget_resx(is_suc: bool) -> Result[str, int]:
+        return Ok("hello") if is_suc else Err(1)
+
+    async def aget_resy(is_suc: bool) -> Result[bool, int]:
+        return Ok(True) if is_suc else Err(2)
+
+    def get_resz(is_suc: bool) -> Result[float, int]:
+        return Ok(0.5) if is_suc else Err(3)
+
+    async def process_xyz(x: str, y: bool, z: float) -> Result[float, int]:
+        return Ok(len(x) + int(y) + z)
+
+    async def _aget_output(
+        is_suc1: bool, is_suc2: bool, is_suc3: bool
+    ) -> Result[float, int]:
+        out: Result[float, int] = await do_async(
+            Ok(w)
+            for x in await aget_resx(is_suc1)
+            for y in await aget_resy(is_suc2)
+            for z in get_resz(is_suc3)
+            for w in await process_xyz(x, y, z)
+        )
+        return out
+
+    assert await _aget_output(True, True, True) == Ok(6.5)
+    assert await _aget_output(True, False, True) == Err(2)
+    assert await _aget_output(False, True, True) == Err(1)
+    assert await _aget_output(False, False, True) == Err(1)
+
+    assert await _aget_output(True, True, False) == Err(3)
+    assert await _aget_output(True, False, False) == Err(2)
+    assert await _aget_output(False, True, False) == Err(1)
+    assert await _aget_output(False, False, False) == Err(1)
+
+
+@pytest.mark.asyncio
+async def test_result_do_general_with_async_values_inline_error() -> None:
+    """
+    Due to subtle behavior, `do()` works in certain cases involving async
+    calls but not others. We surface a more helpful error to the user
+    in cases where it doesn't work indicating to use `do_async()` instead.
+    Contrast this with `test_result_do_general_with_async_values()`
+    in which using `do()` works with async functions as long as
+    their return values are resolved outside the `do()` expression.
+    """
+
+    async def aget_resx(is_suc: bool) -> Result[str, int]:
+        return Ok("hello") if is_suc else Err(1)
+
+    async def aget_resy(is_suc: bool) -> Result[bool, int]:
+        return Ok(True) if is_suc else Err(2)
+
+    def get_resz(is_suc: bool) -> Result[float, int]:
+        return Ok(0.5) if is_suc else Err(3)
+
+    with pytest.raises(TypeError) as excinfo:
+        do(
+            Ok(len(x) + int(y) + z)
+            for x in await aget_resx(True)
+            for y in await aget_resy(True)
+            for z in get_resz(True)
```

Comment:
Random question, does it at all depend on the order in which we have async and non-async lines? For example,
```
for x in foo(...)
for y in await bar(...)
```
vs
```
for x in foo(...)
for y in await bar(...)
```

I don't think it would, but perhaps good to verify in case it's some edge case there as well.